src/Confluent.Kafka/Serialization/FloatDeserializer.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Copyright 2016-2017 Confluent Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Refer to LICENSE for more information.
+
+using System;
+
+namespace Confluent.Kafka.Serialization
+{
+    /// <summary>
+    ///     A deserializer for big endian encoded (network byte ordered) System.Single values.
+    /// </summary>
+    public class FloatDeserializer : IDeserializer<float>
+    {
+        /// <summary>
+        ///     Deserializes a big endian encoded (network byte ordered) System.Single value from a byte array.
+        /// </summary>
+        /// <param name="data">
+        ///     A byte array containing the serialized System.Single value (big endian encoding)
+        /// </param>
+        /// <returns>
+        ///     The deserialized System.Single value.
+        /// </returns>
+        public float Deserialize(byte[] data)
+        {
+            if (data == null)
+            {
+                throw new ArgumentException($"Arg [{nameof(data)}] is null");
+            }
+
+            if (data.Length != 4)
+            {
+                throw new ArgumentException($"Size of {nameof(data)} received by {nameof(FloatDeserializer)} is not 4");
+            }
+
+            float result = default(float);
+            // network byte order -> big endian -> most significant byte in the smallest address.
+            unsafe
+            {
```

Comment:
yes, this needs careful checking! I think I would be happiest to see BitConverter used to check whether we are big or little endian, then the deserialization spelled out in unsafe blocks (one for the little endian case, one for big endian) using pointers as you have it. Same for serialization. On the wire, we should always be big endian.